compiler/src/dotty/tools/dotc/transform/PostTyper.scala

```diff
@@ -534,5 +538,30 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
         sym.addAnnotation(Annotation(defn.ExperimentalAnnot, sym.span))
         sym.companionModule.addAnnotation(Annotation(defn.ExperimentalAnnot, sym.span))
 
+    private def scala2LibPatch(tree: TypeDef)(using Context) =
+      val sym = tree.symbol
+      if compilingScala2StdLib
+        && sym.is(ModuleClass) && !sym.derivesFrom(defn.SerializableClass)
+        && sym.companionClass.derivesFrom(defn.SerializableClass)
+      then
+        // Add Serializable to companion objects of serializable classes
+        tree.rhs match
+          case impl: Template =>
+            val parents1 = impl.parents :+ TypeTree(defn.SerializableType)
+            val impl1 = cpy.Template(impl)(parents = parents1)
+            cpy.TypeDef(tree)(rhs = impl1)
+      else tree
   }
+
+  protected override def infoMayChange(sym: Symbol)(using Context): Boolean =
+    ctx.settings.Yscala2Stdlib.value && sym.isAllOf(ModuleClass, butNot = Package)
```

Comment:
Here you _do_ need the cache since `infoMayChange` is much hotter than the other code where you check for `compilingScala2StdLib`. But moving `compilingScala2StdLib` to be a field of `PostTyper` is sufficient. The problem is that `PostTyper` does not take a context. So you probably need something like this:
```scala
private var myCompilingScala2StdLib: Boolean = false
private var compilingScala2StdLibKnown = false
def compilingScala2StdLib(using Context): Boolean = 
  if !compilingScala2StdLibKnown then 
     myCompilingScala2StdLib = ctx.settings.Yscala2Stdlib.value
     compilingScala2StdLibKnown = true
  myCompilingScala2StdLib
```

Reply:
Ah, I forgot about `initContext`! Agreed, this is the best solution.